### PR TITLE
Add quark script case for CWE 319

### PR DESCRIPTION
**Pull request by zinwang**
Detect CWE-319 in Android Application (ovaa.apk)
==

This scenario seeks to find **the Cleartext Transmission of Sensitive Information**. See [CWE-319](https://cwe.mitre.org/data/definitions/319.html) for more details.

Let's use this [APK](https://github.com/oversecured/ovaa) and the above APIs to show how the Quark script finds this vulnerability. This sample uses the package Retrofit to request Web APIs, but the APIs use cleartext protocols. 

We first design a detection rule ``setRetrofitBaseUrl.json`` to spot on behavior that sets the base URL of the Retrofit instance. Then, we loop through a custom list of cleartext protocol schemes and use API `behaviorInstance.hasString` to filter arguments that are URL strings with cleartext protocol.

Quark Script CWE-319.py
--

``` python    
from quark.script import runQuarkAnalysis, Rule

SAMPLE_PATH = "./ovaa.apk"
RULE_PATH = "setRetrofitBaseUrl.json"

PROTOCOL_KEYWORDS = [
    "http",
    "smtp",
    "ftp"
]


ruleInstance = Rule(RULE_PATH)
quarkResult = runQuarkAnalysis(SAMPLE_PATH, ruleInstance)

for setRetrofitBaseUrl in quarkResult.behaviorOccurList: 
    for protocol in PROTOCOL_KEYWORDS:
        
        regexRule = f"{protocol}://[0-9A-Za-z./-]+"
        cleartextProtocolUrl = setRetrofitBaseUrl.hasString(regexRule, True)
        
        if cleartextProtocolUrl:
            print(f"CWE-319 detected!")
            print(f"Here are the found URLs with cleartext protocol:")
            print("\n".join(cleartextProtocolUrl))
```

Quark Rule: setRetrofitBaseUrl.json
--

``` json
{
    "crime": "Set Retrofit Base Url",
    "permission": [],
    "api": 
    [
        {
            "descriptor": "()V",
            "class": "Lretrofit2/Retrofit$Builder;",
            "method": "<init>"
        },
        {
            "descriptor": "(Ljava/lang/String;)Lretrofit2/Retrofit$Builder;",
            "class": "Lretrofit2/Retrofit$Builder;",
            "method": "baseUrl"
        }
    ],
    "score": 1,
    "label": []
}
```

Quark Script Result
--

```
$ python3 CWE-319.py
CWE-319 detected!
Here are the found URLs with cleartext protocol:
http://example.com./api/v1/
```

### CODE REVIEW

**Comment by 18z** [comment]
HI @zinwang

I found an indention error.

WRONG
```python
for setRetrofitBaseUrl in quarkResult.behaviorOccurList: 
                
urls = setRetrofitBaseUrl.hasUrl()
```
CORRECT
```python
for setRetrofitBaseUrl in quarkResult.behaviorOccurList: 
        urls = setRetrofitBaseUrl.hasUrl()
```

**Comment by 18z** [comment]
Another indention error?

WRONG
```python
for url, protocol in product(urls, PROTOCOL_KEYWORDS):
    if url.startswith(protocol):
    print(f"CWE-319 detected! URL using cleartext protocol, \"{url}\" found!")
```

CORRECT
```python
for url, protocol in product(urls, PROTOCOL_KEYWORDS):
    if url.startswith(protocol):
        print(f"CWE-319 detected! URL using cleartext protocol, \"{url}\" found!")
```

**Comment by 18z** [comment]
One question:
What if the URL is https://www.google.com?

```python
for url, protocol in product(urls, PROTOCOL_KEYWORDS):
    if url.startswith(protocol):
    print(f"CWE-319 detected! URL using cleartext protocol, \"{url}\" found!")
```

**Comment by pulorsok** [comment]
@zinwang 
I think you are trying to print all the URLs with `PROTOCOL_KEYWORDS` in the APK sample.
Indention error aside, this can only print the last `setRetrofitBaseUrl` in `quarkResult.behaviorOccurList`.
```python
for setRetrofitBaseUrl in quarkResult.behaviorOccurList: 
urls = setRetrofitBaseUrl.hasUrl()
                    
for url, protocol in product(urls, PROTOCOL_KEYWORDS):
    if url.startswith(protocol):
    print(f"CWE-319 detected! URL using cleartext protocol, \"{url}\" found!")
```

Maybe change to:

```python
for setRetrofitBaseUrl in quarkResult.behaviorOccurList: 
    urls = setRetrofitBaseUrl.hasUrl()
                    
    for url, protocol in product(urls, PROTOCOL_KEYWORDS):
        if url.startswith(protocol):
              print(f"CWE-319 detected! URL using cleartext protocol, \"{url}\" found!")
```

**Comment by 18z** [comment]
I found another problem.

```python
behaviorInstance.hasUrl()
```

It seems to me that ```hasUrl()``` can only return ```URL``` that starts with ```http```.
Other protocols such as ```smtp``` and ```ftp``` cannot be returned.

Screen Shot 2022-11-01 at 9 54 14 AM

File: docs/source/quark_script.rst [hunk 1]
```diff
@@ -907,3 +907,73 @@ Quark Script Result
...
+    PROTOCOL_KEYWORDS = [
+        "http",
+        "smtp",
+        "ftp"
+    ]
+
+    ruleInstance = Rule(RULE_PATH)
+    quarkResult = runQuarkAnalysis(SAMPLE_PATH,ruleInstance)
+
+    for setRetrofitBaseUrl in quarkResult.behaviorOccurList:
```
Comment (sidra-asa) [inline docs/source/quark_script.rst:940]: There are indention errors.

File: docs/build/html/genindex.html [hunk 2]
```diff
@@ -198,21 +199,13 @@ <h2 id="B">B</h2>
       <li><a href="quark.core.interface.html#quark.core.interface.baseapkinfo.BaseApkinfo">BaseApkinfo (class in quark.core.interface.baseapkinfo)</a>
 </li>
       <li><a href="quark.script.html#quark.script.Behavior">Behavior (class in quark.script)</a>
-
```
Comment (sidra-asa) [inline docs/build/html/genindex.html:201]: It seems to overwrite the content of previous commits.

**Comment by zinwang** [comment]
Hi @18z @sidra-asa @pulorsok
 
I've updated the PR.
Here is what I have done.

1. Rewrite the Quark Script, using regex to match protocol keywords. It solves [this problem](https://github.com/quark-engine/quark-engine/pull/413#issuecomment-1297871089).

2. Rewrite the Quark Script, using hasString instead of hasUrl API. It solves [this problem](https://github.com/quark-engine/quark-engine/pull/413#issuecomment-1297912174).

3. Make sure the commits are correct. Solve [this problem](https://github.com/quark-engine/quark-engine/pull/413#discussion_r1010183424).

4. Make sure the indents are correct.

If there is any problem, please tell me! Thank you!

**Comment by sidra-asa** [comment]
@zinwang 

Here's what I found:

* content of docs/build/html/genindex.html is overwritten, compared to previous commits.
Please confirm whether it is proper.

* I followed the instructions, but the result is not the same.
Did I miss something? 
Screen Shot 2022-11-02 at 1 30 50 AM

P.S. I put the message about quarkResult.behaviorOccurList for debugging.

**Comment by 18z** [comment]
LGTM, @18z approved this PR.

**Comment by haeter525** [comment]
Hi @zinwang. I found two lines that don't follow the PEP 8 style.

```python
# line 14
quarkResult = runQuarkAnalysis(SAMPLE_PATH,ruleInstance)
```
```python
# line 20
        cleartextProtocolUrl = setRetrofitBaseUrl.hasString(regexRule,True)
```
There should be one whitespace between `,` and the second argument.

**Review by 18z — APPROVED** [review]

**Comment by sidra-asa** [comment]
> @zinwang
> 
> Here's what I found:
> 
> * content of docs/build/html/genindex.html is overwritten, compared to previous commits.
>   Please confirm whether it is proper.
> * I followed the instructions, but the result is not the same.
>   Did I miss something?

@zinwang 

I just realized that we are using different ovaa.apk samples.
I used to test the sample which is downloaded from the [issue](https://github.com/oversecured/ovaa/issues/3). 
It's my fault.

Now, I test the sample both from you and from the [official repo](https://github.com/dark-warlord14/ovaa/releases/tag/1.0), and the results are all fine.
Thanks for your help.

**Comment by zinwang** [comment]
@sidra-asa @haeter525 

I've fixed the overwriting problem in the document and made sure the codes meet the PEP8 format.
Please have a look.

**Review by sidra-asa — APPROVED** [review]
LGTM

**Pull request merged**